File: Data Structures and Problem Solving Using Java 4th/Xref.java
import java.io.InputStreamReader;
import java.io.IOException;
import java.io.FileReader;
import java.io.Reader;

import weiss.util.Iterator;
import weiss.util.TreeMap;
import weiss.util.List;
import weiss.util.ArrayList;
import weiss.util.Map;

// Xref class interface: generate cross-reference
//
// CONSTRUCTION: with a Reader object
// ******************PUBLIC OPERATIONS***********************
// void generateCrossReference( ) --> Name says it all ...
// ******************ERRORS**********************************
// Error checking on comments and quotes is performed
// main generates cross-reference.


/**
 * Class to perform cross reference
 * generation for Java programs.
 */
public class Xref
{
    /**
     * Constructor.
     * @param inStream the stream containing a program.
     */
    public Xref( Reader inStream )
    {
        tok = new Tokenizer( inStream );
    }


    private Tokenizer tok;   // tokenizer object


    /**
     * Output the cross reference.
     */
    public void generateCrossReference( )
    {
        Map<String,List<Integer>> theIdentifiers = new TreeMap<String,List<Integer>>( );
        String current;

            // Insert identifiers into the search tree
        while( ( current = tok.getNextID( ) ) != "" )
        {
            List<Integer> lines = theIdentifiers.get( current );
            if( lines == null )
            {
                lines = new ArrayList<Integer>( );
                theIdentifiers.put( current, lines );
            }
            lines.add( tok.getLineNumber( ) );
        }
        
            // Iterate through search tree and output
            // identifiers and their line number
			
        for( Map.Entry<String,List<Integer>> thisNode : theIdentifiers.entrySet( ) )
        {
            Iterator<Integer> lineItr = thisNode.getValue( ).iterator( );

                // Print identifier and first line where it occurs
            System.out.print( thisNode.getKey( ) + ": " );
            System.out.print( lineItr.next( ) );
            
                // Print all other lines on which it occurs
            while( lineItr.hasNext( ) )
                System.out.print( ", " + lineItr.next( ) );
            System.out.println( );    
        }    
    }

    /**
     * main routine to generate cross-reference.
     * If no command line parameters, standard input is used.
     * Otherwise, files in command line are used.
     */
    public static void main( String [ ] args )
    {

        Xref p;

        if( args.length == 0 )
        {
            p = new Xref( new InputStreamReader( System.in ) );
            p.generateCrossReference( );

            return;
        }

        for( int i = 0; i < args.length; i++ )
        {
            FileReader f = null;
            try 
            {
                f = new FileReader( args[ i ] );

                System.out.println( args[ i ] + ": " );
                p = new Xref( f );
                p.generateCrossReference( );
                f.close( );
            }
            catch( IOException e )
            {
                System.err.println( e + args[ i ] );
            }
            finally
            {
                try
                {
                    if( f != null )
                        f.close( );
                }
                catch( IOException e )
                {
                }
            }            
        }
    }
}

File: Data Structures and Problem Solving Using Cpp 2nd/Xref.cpp
// Xref class interface: generate cross-reference.
//
// CONSTRUCTION: with an istream object.
//
// ******************PUBLIC OPERATIONS***********************
// void generateCrossReference( ) --> Name says it all ...
// ******************ERRORS**********************************
// Error checking on comments and quotes is performed.

#include "Tokenizer.h"

#ifdef USE_DOT_H
    #include <iostream.h>
    #include <fstream.h>
#else
    #include <fstream>
    #include <iostream>
    #if !defined( __BORLANDC__ ) || __BORLANDC__ >= 0x0530
        using namespace std;      // Borland 5.0 has a bug
    #endif
#endif
    
#ifndef SAFE_STL
    #include <list>
    #include <string>
    #include <map>
    #include <functional>
    using namespace std;
#else
    #include "list.h"
    #include "mystring.h"
    #include "map.h"
    #include "functional.h"
    #include "StartConv.h"
#endif


class Xref
{
  public:
    Xref( istream & input ) : tok( input ) { }

    void generateCrossReference( );

  private:
    Tokenizer tok;      // Token source
};

// Output the cross reference.
void Xref::generateCrossReference( )
{
    typedef map<string,list<int>, less<string> > MapType;
    typedef MapType::const_iterator MapIteratorType;

    MapType xrefMap;
    string current;

      // Insert identifiers into the map.
    while( ( current = tok.getNextID( ) ) != "" )
        xrefMap[ current ].push_back( tok.getLineNumber( ) );

      // Iterate through map and output
      // identifiers and their line number.
    MapIteratorType itr;
    for( itr = xrefMap.begin( ); itr != xrefMap.end( ); ++itr )
    {
        const list<int> & theLines = (*itr).second;
        list<int>::const_iterator lineItr = theLines.begin( );

          // Print identifier and first line where it occurs.
        cout << (*itr).first << ": " << *lineItr;


          // Print all other lines on which it occurs.
        for( ++lineItr; lineItr != theLines.end( ); ++lineItr )
            cout << ", " << *lineItr;
        cout << endl;
    }
}


int main( int argc, const char **argv )
{
    if( argc == 1 )
    {
        Xref p( cin );
        p.generateCrossReference( );
        return 0;
    }

    while( --argc )
    {
        ifstream ifp( *++argv );
        if( !ifp )
        {
            cout << "Cannot open " << *argv << endl;
            continue;
        }
        cout << "*********" << *argv << "*********\n";
        Xref p( ifp );
        p.generateCrossReference( );
    }

    return 0;
}

#ifdef SAFE_STL
    #include "EndConv.h"
#endif
